man/halt.8
.\"{{{}}}
.\"{{{  Title
.TH HALT 8 "Nov 6, 2001" "" "Linux System Administrator's Manual"
.\"}}}
.\"{{{  Name
.SH NAME
halt, reboot, poweroff \- stop the system.
.\"}}}
.\"{{{  Synopsis
.SH SYNOPSIS
.B /sbin/halt
.RB [ \-n ]
.RB [ \-w ]
.RB [ \-d ]
.RB [ \-f ]
.RB [ \-i ]
.RB [ \-p ]
.RB [ \-h ]
.br
.B /sbin/reboot
.RB [ \-n ]
.RB [ \-w ]
.RB [ \-d ]
.RB [ \-f ]
.RB [ \-i ]
.br
.B /sbin/poweroff
.RB [ \-n ]
.RB [ \-w ]
.RB [ \-d ]
.RB [ \-f ]
.RB [ \-i ]
.RB [ \-h ]
.\"}}}
.\"{{{  Description
.SH DESCRIPTION
\fBHalt\fP notes that the system is being brought down in the file
\fI/var/log/wtmp\fP, and then either tells the kernel to halt, reboot or
power-off the system.
.PP
If \fBhalt\fP or \fBreboot\fP is called when the system is
\fInot\fP in runlevel \fB0\fP or \fB6\fP, in other words when it's running
normally, \fBshutdown\fP will be invoked instead (with the \fB-h\fP
or \fB-r\fP flag). For more info see the \fBshutdown\fP(8)
manpage.
.PP
The rest of this manpage describes the behaviour in runlevels 0
and 6, that is when the systems shutdown scripts are being run.
.\"}}}
.\"{{{  Options
.SH OPTIONS
.IP \fB\-n\fP
Don't sync before reboot or halt. Note that the kernel and storage
drivers may still sync.
.IP \fB\-w\fP
Don't actually reboot or halt but only write the wtmp record
(in the \fI/var/log/wtmp\fP file).
.IP \fB\-d\fP
Don't write the wtmp record. The \fB\-n\fP flag implies \fB\-d\fP.
.IP \fB\-f\fP
Force halt or reboot, don't call \fBshutdown\fP(8).
.IP \fB\-i\fP
Shut down all network interfaces just before halt or reboot.
.IP \fB\-h\fP
Put all hard drives on the system in stand-by mode just before halt or power-off.
.IP \fB\-p\fP
When halting the system, switch off the power. This is the default when halt is
called as \fBpoweroff\fP.
.\"}}}
.\"{{{  Diagnostics
.SH DIAGNOSTICS
If you're not the superuser, you will get the message `must be superuser'.
.\"}}}
.\"{{{  Notes
.SH NOTES
Under older \fBsysvinit\fP releases , \fBreboot\fP and \fBhalt\fP should
never be called directly. From release 2.74 on \fBhalt\fP and \fBreboot\fP
invoke \fBshutdown\fP(8) if the system is not in runlevel 0 or 6. This
means that if \fBhalt\fP or \fBreboot\fP cannot find out the current
runlevel (for example, when \fI/var/run/utmp\fP hasn't been initialized
correctly) \fBshutdown\fP will be called, which might not be what you want.
Use the \fB-f\fP flag if you want to do a hard \fBhalt\fP or \fBreboot\fP.
.PP
The \fB-h\fP flag puts all hard disks in standby mode just before halt
or power-off. Right now this is only implemented for IDE drives. A side
effect of putting the drive in stand-by mode is that the write cache
on the disk is flushed. This is important for IDE drives, since the
kernel doesn't flush the write cache itself before power-off.
.PP
The \fBhalt\fP program uses /proc/ide/hd* to find all IDE disk devices,
which means that \fI/proc\fP needs to be mounted when \fBhalt\fP or
\fBpoweroff\fP is called or the \fB-h\fP switch will do nothing.
.PP
.\"}}}
.\"{{{  Author
.SH AUTHOR
Miquel van Smoorenburg, miquels@cistron.nl
.\"}}}
.\"{{{  See also
.SH "SEE ALSO"
.BR shutdown (8),
.BR init (8)
.\"}}}


debian/changelog
sysvinit (2.87dsf-8slp2+s4) unstable; urgency=low

  * Bring sysvinit's postinst script back

 -- Karol Lewandowski <k.lewandowsk@samsung.com>  Thu, 01 Jul 2010 11:23:54 +0200

sysvinit (2.87dsf-8slp2+s3) unstable; urgency=low

  * SLP: exclude sysvinit from dh_installdeb to remove postinst script
    from that package.

 -- Rafal Krypa <r.krypa@samsung.com>  Thu, 10 Jun 2010 17:06:58 +0200

sysvinit (2.87dsf-8slp2+s2) unstable; urgency=low

  * Revert "src/Makefile: workaround for Scratchbox, always use -lcrypt
    for sulogin."
  * Revert "Decrease debhelper compatibility level to 5 to allow
    building under Scratchbox."
  * debian/rules: rework "drop selinux dependency"
  * Set myself as maintainer.

 -- Rafal Krypa <r.krypa@samsung.com>  Tue, 20 Apr 2010 16:59:19 +0200

sysvinit (2.87dsf-8+s1) unstable; urgency=low

  * Remove forced root filesystem check from etc/init.d/checkroot.sh -
    it is against fstab semantics and triggered a bug in ubifs.
  * getty settings in inittab appropriate for SLP.
  * Relax dependencies on mount and coreutils for initscripts to allow
    busybox replacements.
  * Drop selinux dependency.
  * Decrease debhelper compatibility level to 5 to allow building under
    Scratchbox.
  * debian/control: replace the Uploaders field.
  * Adjust initscripts to for the SLP platform.
  * src/Makefile: workaround for Scratchbox, always use -lcrypt for
    sulogin.

 -- Rafal Krypa <r.krypa@samsung.com>  Wed, 03 Feb 2010 14:48:02 +0900

sysvinit (2.87dsf-8) unstable; urgency=low

  * Update patch 63_init_keep_utf8_ttyflag.patch to try to get it
    working on freebsd and hurd too.
  * Debconf templates and debian/control changed based on review
    by the debian-l10n- english team as part of the Smith review
    project (Closes: #549539).  Thanks to Christian Perrier for
    the patch.
  * Added section in README.Debian for sysv-rc, explaining the
    migration in more detail.  Thanks to Justin B Rye for the text.
  * Debconf translations:
    - Added Vietnamese from Clytie Siddall (Closes: #550220).
    - Added Swedish from Martin Ågren (Closes: #550495).
    - Added Czech from Miroslav Kure (Closes: #551325).
    - Added Spanish from Francisco Javier Cuadrado (Closes: #551594).
    - Added Portuguese from António Moreira (Closes: #551675).
    - Added French from Steve Petruzzello (Closes: #551722).
    - Added Finnish from Esko Arajärvi (Closes: #551916).
    - Added Italian from Luca Monducci (Closes: #551922).

 -- Petter Reinholdtsen <pere@debian.org>  Sun, 25 Oct 2009 21:49:44 +0100

sysvinit (2.87dsf-7) unstable; urgency=low

  [ Petter Reinholdtsen ]
  * Change if-up.d/mountnfs to not claim to wait for the last network
    interface when there are no network file systems listed in
    /etc/fstab (Closes: #512237, 481028).  Patch from Adrian Bridgett.
  * Add startpar patch 05_pri_kdm_gdm.patch to try to get gdm and kdm
    to start earlier when concurrent booting is enabled.
  * Quiet down sysv-rc postinst to not give error when no file exist
    in /var/lib/update-rc.d.  Discovered from piuparts.
  * Report error from update-rc.d when insserv reject a script, to
    make it more clear what program call failed in a postinst.
  * Drop optional dependency from checkfs to cryptdisks, the
    cryptdisks script have a reverse dependency on checkfs, and that
    is a better way to do it.
  * Rewrite init.d/mtab.sh and init.d/checkroot.sh to not use
    /usr/bin/which, as it is running before /usr/ is guaranteed to be
    available.
  * Rewrite /lib/init/usplash-fsck-functions.sh to use blkid instead of
    the now obsolete vol_id.  Patch from Martin Pitt and Ubuntu.
  * Update 91_sulogin_lockedpw.dpatch to include documentation,
    without the Ubuntu references.  Based on patch from Scott James
    Remnant and Ubuntu.
  * Add patch 55_bootlogd_flush.patch to make sure lines are flushed
    to the kernel even when not asking the kernel to flush to disk
    (Closes: 542515).  Patch from Scott Gifford.
  * Add patch 63_init_keep_utf8_ttyflag.patch to make sure the utf-8
    tty flag is not cleared (Closes: 547073).  Patch from Samuel
    Thibault.
  * Include url to wiki page with information on how to fix the
    problems when unable to migrate to dependency based boot
    sequencing.
  * Make mountnfs-bootclean.sh also depend on $local_fs, to avoid
    surprises on systems where the admin removed mountnfs.sh from
    rcS.d/.  It is not a good idea to remove it, but there is no need
    for the script to fail when someone is crazy enough to remove a
    essential init.d script

  [ Henrique de Moraes Holschuh ]
  * invoke-rc.d: return exit status 4 when action "status" is denied,
    to allow simpleminded policy-rc.d scripts to work well with the
    status action (closes: #381497)

  [ Kel Modderman ]
  * Fix bootlogs init.d script to output correct name and supported
    actions in usage statement. (Closes: #551263)

 -- Petter Reinholdtsen <pere@debian.org>  Sun, 25 Oct 2009 11:07:22 +0100

sysvinit (2.87dsf-6) unstable; urgency=low

  * Correct code in sysv-rc postinst to avoid failing to install when no
    init.d scripts have been removed (Closes: #546405).
  * Start bootlogd before lvm2 too, to push it even further forward
    in the boot sequence.

 -- Petter Reinholdtsen <pere@debian.org>  Tue, 15 Sep 2009 01:28:19 +0200

sysvinit (2.87dsf-5) experimental; urgency=low

  * Uploading to experimental, to test the new build rules.

  * Make sysv-rc postinst report detected problems to stderr too when
    failing to migrate.
  * Fix typo in error message from postinst (Closes: #545409).
  * Make initscripts depend on sysvinit-utils (>= 2.86.ds1-64), to
    make sure the fstab-decode program is available (Closes: #545356).
  * Make sure the calls to 'update-rc.d X remove' in initscripts
    postinst do not ignore errors (Closes: #406361).
  * Make sysvinit depend on sysvinit-utils (>= 2.86.ds1-66) to avoid

src/halt.c
/*
 * Halt		Stop the system running.
 *		It re-enables CTRL-ALT-DEL, so that a hard reboot can
 *		be done. If called as reboot, it will reboot the system.
 *
 *		If the system is not in runlevel 0 or 6, halt will just
 *		execute a "shutdown -h" to halt the system, and reboot will
 *		execute an "shutdown -r". This is for compatibility with
 *		sysvinit 2.4.
 *
 * Usage:	halt [-n] [-w] [-d] [-f] [-h] [-i] [-p]
 *		-n: don't sync before halting the system
 *		-w: only write a wtmp reboot record and exit.
 *		-d: don't write a wtmp record.
 *		-f: force halt/reboot, don't call shutdown.
 *		-h: put harddisks in standby mode
 *		-i: shut down all network interfaces.
 *		-p: power down the system (if possible, otherwise halt).
 *
 *		Reboot and halt are both this program. Reboot
 *		is just a link to halt. Invoking the program
 *		as poweroff implies the -p option.
 *
 * Author:	Miquel van Smoorenburg, miquels@cistron.nl
 *
 * Version:	2.86,  30-Jul-2004
 *
 *		This file is part of the sysvinit suite,
 *		Copyright 1991-2004 Miquel van Smoorenburg.
 *
 *		This program is free software; you can redistribute it and/or
 *		modify it under the terms of the GNU General Public License
 *		as published by the Free Software Foundation; either version
 *		2 of the License, or (at your option) any later version.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <stdlib.h>
#include <utmp.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <sys/times.h>
#include <time.h>
#include <signal.h>
#include <stdio.h>
#include <getopt.h>
#include "reboot.h"

char *Version = "@(#)halt  2.86  31-Jul-2004 miquels@cistron.nl";
char *progname;

#define KERNEL_MONITOR	1 /* If halt() puts you into the kernel monitor. */
#define RUNLVL_PICKY	0 /* Be picky about the runlevel */

extern int ifdown(void);
extern int hddown(void);
extern void write_wtmp(char *user, char *id, int pid, int type, char *line);

/*
 *	Send usage message.
 */
void usage(void)
{
	fprintf(stderr, "usage: %s [-n] [-w] [-d] [-f] [-h] [-i]%s\n",
		progname, strcmp(progname, "halt") ? "" : " [-p]");
	fprintf(stderr, "\t-n: don't sync before halting the system\n");
	fprintf(stderr, "\t-w: only write a wtmp reboot record and exit.\n");
	fprintf(stderr, "\t-d: don't write a wtmp record.\n");
	fprintf(stderr, "\t-f: force halt/reboot, don't call shutdown.\n");
	fprintf(stderr, "\t-h: put harddisks in standby mode.\n");
	fprintf(stderr, "\t-i: shut down all network interfaces.\n");
	if (!strcmp(progname, "halt"))
		fprintf(stderr, "\t-p: power down the system (if possible, otherwise halt).\n");
	exit(1);
}

/*
 *	See if we were started directly from init.
 *	Get the runlevel from /var/run/utmp or the environment.
 */
int get_runlevel(void)
{
	struct utmp *ut;
	char *r;
#if RUNLVL_PICKY
	time_t boottime;
#endif

	/*
	 *	First see if we were started directly from init.
	 */
	if (getenv("INIT_VERSION") && (r = getenv("RUNLEVEL")) != NULL)
		return *r;

	/*
	 *	Hmm, failed - read runlevel from /var/run/utmp..
	 */
#if RUNLVL_PICKY
	/*
	 *	Get boottime from the kernel.
	 */
	time(&boottime);
	boottime -= (times(NULL) / HZ);
#endif

	/*
	 *	Find runlevel in utmp.
	 */
	setutent();
	while ((ut = getutent()) != NULL) {
#if RUNLVL_PICKY
		/*
		 *	Only accept value if it's from after boottime.
		 */
		if (ut->ut_type == RUN_LVL && ut->ut_time > boottime)
			return (ut->ut_pid & 255);
#else
		if (ut->ut_type == RUN_LVL)
			return (ut->ut_pid & 255);
#endif
	}
	endutent();

	/* This should not happen but warn the user! */
	fprintf(stderr, "WARNING: could not determine runlevel"
		" - doing soft %s\n", progname);
	fprintf(stderr, "  (it's better to use shutdown instead of %s"
		" from the command line)\n", progname);

	return -1;
}

/*
 *	Switch to another runlevel.
 */
void do_shutdown(char *fl, char *tm)
{
	char *args[8];
	int i = 0;

	args[i++] = "shutdown";
	args[i++] = fl;
	if (tm) {
		args[i++] = "-t";
		args[i++] = tm;
	}
	args[i++] = "now";
	args[i++] = NULL;

	execv("/sbin/shutdown", args);
	execv("/etc/shutdown", args);
	execv("/bin/shutdown", args);

	perror("shutdown");
	exit(1);
}

/*
 *	Main program.
 *	Write a wtmp entry and reboot cq. halt.
 */
int main(int argc, char **argv)
{
	int do_reboot = 0;
	int do_sync = 1;
	int do_wtmp = 1;
	int do_nothing = 0;
	int do_hard = 0;
	int do_ifdown = 0;
	int do_hddown = 0;
	int do_poweroff = 0;
	int c;
	char *tm = NULL;

	/*
	 *	Find out who we are
	 */
	/* Remove dash passed on in argv[0] when used as login shell. */
	if (argv[0][0] == '-') argv[0]++;
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	if (!strcmp(progname, "reboot")) do_reboot = 1;
	if (!strcmp(progname, "poweroff")) do_poweroff = 1;

	/*
	 *	Get flags
	 */
	while((c = getopt(argc, argv, ":ihdfnpwt:")) != EOF) {
		switch(c) {
			case 'n':
				do_sync = 0;
				do_wtmp = 0;
				break;
			case 'w':
				do_nothing = 1;
				break;
			case 'd':
				do_wtmp = 0;
				break;
			case 'f':
				do_hard = 1;
				break;
			case 'i':
				do_ifdown = 1;
				break;
			case 'h':
				do_hddown = 1;
				break;
			case 'p':
				do_poweroff = 1;
				break;
			case 't':
				tm = optarg;
				break;
			default:
				usage();
		}
	 }
	if (argc != optind) usage();

	if (geteuid() != 0) {
		fprintf(stderr, "%s: must be superuser.\n", progname);
		exit(1);
	}

	(void)chdir("/");

	if (!do_hard && !do_nothing) {
		/*
		 *	See if we are in runlevel 0 or 6.
		 */
		c = get_runlevel();
		if (c != '0' && c != '6')
			do_shutdown(do_reboot ? "-r" : "-h", tm);
	}

	/*
	 *	Record the fact that we're going down
	 */
	if (do_wtmp)
		write_wtmp("shutdown", "~~", 0, RUN_LVL, "~~");

	/*
	 *	Exit if all we wanted to do was write a wtmp record.
	 */
	if (do_nothing && !do_hddown && !do_ifdown) exit(0);

	if (do_sync) {
		sync();
		sleep(2);
	}

	if (do_ifdown)
		(void)ifdown();

	if (do_hddown)
		(void)hddown();

	if (do_nothing) exit(0);

	if (do_reboot) {
		init_reboot(BMAGIC_REBOOT);
	} else {
		/*
		 *	Turn on hard reboot, CTRL-ALT-DEL will reboot now
		 */
#ifdef BMAGIC_HARD
		init_reboot(BMAGIC_HARD);
#endif

		/*
		 *	Stop init; it is insensitive to the signals sent
		 *	by the kernel.
		 */
		kill(1, SIGTSTP);

		/*
		 *	Halt or poweroff.
		 */
		if (do_poweroff)
			init_reboot(BMAGIC_POWEROFF);
		/*
		 *	Fallthrough if failed.
		 */
		init_reboot(BMAGIC_HALT);
	}

	/*
	 *	If we return, we (c)ontinued from the kernel monitor.
	 */
#ifdef BMAGIC_SOFT
	init_reboot(BMAGIC_SOFT);
#endif
	kill(1, SIGCONT);

	exit(0);
}
